include/boost/compute/system.hpp

```diff
@@ -280,6 +295,145 @@ class system
     {
         return str.find(pattern) != std::string::npos;
     }
+
+    /// \internal_
+    static device init_default_device(device* user_device = 0)
+    {
+        static device default_device;
+#ifdef BOOST_COMPUTE_THREAD_SAFE
+    #ifdef BOOST_COMPUTE_USE_CPP11
+        using namespace std;
+    #else 
+        using namespace boost;
+    #endif 
+        static atomic<bool> is_init;
+        static mutex init_mutex;
+
+        bool is_init_value = is_init.load(memory_order_consume);
+        if (!is_init_value)
+        {
+            lock_guard<mutex> lock(init_mutex);
+            is_init_value = is_init.load(memory_order_consume);
+            if (!is_init_value)
+            {
+                default_device = user_device ? 
+                    *user_device : find_default_device();
+
+                is_init.store(true, memory_order_release);
+            }
+        }
+#else
+        if (!default_device.get())
+        {
+            default_device = user_device ? 
+                *user_device : find_default_device();
+        }
+#endif       
+        return default_device;
+    }
+
+    /// \internal_
+    static context init_default_context(context* user_context = 0)
+    {
+        static context default_context;
+
+#ifdef BOOST_COMPUTE_THREAD_SAFE
+    #ifdef BOOST_COMPUTE_USE_CPP11
+        using namespace std;
+    #else 
+        using namespace boost;
+    #endif 
+        static atomic<bool> is_init;
+        static mutex init_mutex;
+
+        bool is_init_value = is_init.load(memory_order_consume);
+        if (!is_init_value)
+        {
+            lock_guard<mutex> lock(init_mutex);
+            is_init_value = is_init.load(memory_order_consume);
+            if (!is_init_value)
+            {
+                default_context = user_context ? 
+                    *user_context : context(default_device());
+
+                is_init.store(true, memory_order_release);
+            }
+        }
+#else // BOOST_COMPUTE_THREAD_SAFE
+        if (!default_context.get())
+        {
+            default_context = user_context ?
+                *user_context : context(default_device());
+        }
+#endif // BOOST_COMPUTE_THREAD_SAFE
+        return default_context;
+    }
+
+    /// \internal_
+    static void init_default_device_and_context(command_queue* user_queue)
+    {
+        device user_device = user_queue->get_device();
+        context user_context = user_queue->get_context();
+
+        if ( (user_device != init_default_device(&user_device)) ||
+             (user_context != init_default_context(&user_context)) )
+        {
+            // Try invoking default_queue() before anything else
+            BOOST_THROW_EXCEPTION(context_error(&user_context, 
```

Comment:
I'm not sure if `context_error` is right error here.  I think it should be custom exception similar in design to `no_device_found`.

Reply:
Glad you pointed that out. The original design was to set user-provided context so it sort of fits in. But now it's been user-provided command queue already. I think we can throw a more explicit exception `set_default_queue_error`

Reply:
The name of the error sounds ok.